v5 release post (#93)
chainer/chainer.org: _posts/2018-10-25-released-v5.md

---
title: Released Chainer/CuPy v5.0.0
layout: post
categories: Announcement
author: Seiya Tokui
---

We have released Chainer and CuPy v5.0.0 today!
This is a major release that introduces several new features.

The following is a list of selected updates. Full updates can be found in the release notes: [Chainer](https://github.com/chainer/chainer/releases/tags/v5.0.0), [CuPy](https://github.com/cupy/cupy/releases/tag/v5.0.0).

> - **Static subgraph optimization** *(experimental)*.

Reviewer: ```suggestion
- **Static subgraph optimization** *(experimental)*. (Code name: Uneager Execution)
```

Author: Isn't it confusing? We are not using such a code name and also this feature is not about eagerness as I understand (the first iteration anyway runs in define-by-run style).

>   By applying `@static_graph` decorator to the static part of your computation (which uses the same graph at every iteration), the computational graph of that part is cached and reused.

Reviewer 2: ```suggestion
  By applying the `@static_graph` decorator to the static part of your computation (which uses the same graph at every iteration), the computational graph of that part is cached and reused.
```

>   For fully-static models, we are observing 20-60% speed up in most cases.

Reviewer: ```suggestion
  Fully-static models speed up by 20-60% in most cases.
```

  Example code modified for the static subgraph feature can be found [here](https://github.com/chainer/chainer/tree/v5/examples/static_graph_optimizations).
- **Float16 support**.

>   Using half-precision float is made much easier!

Reviewer: ```suggestion
  Using half-precision floats is made much easier!
```

>   Since recent GPU technologies are focusing on half and mixed precision computations, using float16 is crucial for fully utilizing the latest hardware performance.

Reviewer: ```suggestion
  Since recent GPU technologies often focu on half and mixed precision computations, using float16 is crucial for fully utilizing the latest hardware performance.
```

  In Chainer v5, the default floating point dtype is configurable via `CHAINER_DTYPE` environment variable or `config.dtype` entry.

>   Using this feature, you can start trying float16 without modifying your code in most cases.

Reviewer: ```suggestion
  Using this feature, most code will be able to use float16 without modification.
```

>   Many classes and functions fixed to support float16 inputs and parameters.

Reviewer 2: ```suggestion
  Many classes and functions are fixed to support float16 inputs and parameters.
```

>   Many classes and functions are also fixed to support float16 inputs and parameters.

Reviewer: ```suggestion
  Many classes and functions fixed to support float16 inputs and parameters.
```

- **ChainerMN integration**.

>   [ChainerMN](https://github.com/chainer/chainermn) is an add-on package of Chainer for distributed deep learning.

Reviewer 3: I think it'd better be a past tense like: ChainerMN was an add-on package ..., but is no longer an add-on and has become officially built-in module o Chainer v5.

Author: Looks good. Applied.

>   but is no longer an add-on and has become officially a built-in module of Chainer v5.

Reviewer: ```suggestion
  but now a built-in module of Chainer v5.
```

Author: fixed

>   The APIs and the usage are not changed; you only need to install `chainer` and `mpi4py` to start distributed deep learning.

Reviewer: ```suggestion
  The APIs and the usage are not changed; just install `chainer` and `mpi4py` to start distributed deep learning.
```

- **Probability distributions**.

>   We introduced `chainer.distributions` module that implements many parametric probability distributions with autograd capability.

Reviewer 2: ```suggestion
  We introduced the `chainer.distributions` module that implements many parametric probability distributions with autograd capability.
```

  Each distribution provides point-wise evaluation (e.g. log density), statistics computation, and sampling.
  For its implementation, we also added many GPU sampling routines (under `cupy.random`) and special functions (e.g. log-gamma function).
  While v5 includes many frequently used distributions, we are still expanding this feature for the upcoming releases.
- **iDeep 2.0**.
  Chainer Backend for Intel Architecture, a.k.a. iDeep, is updated.

>   You can install it by `pip install ideep4py`, and use it by setting the environment variable `CHAINER_USE_IDEEP=auto`.

Reviewer: ```suggestion
  You can install it with `pip install ideep4py`, and use it by setting the environment variable `CHAINER_USE_IDEEP=auto`.
```

  There are many performance improvements in this version.
- **CuPy interoperability with other libraries and ecosystems**.

>   CuPy ndarray is now easily combined with other libraries.

Reviewer: ```suggestion
  CuPy ndarray can now be easily combined with other libraries.
```

>   CuPy ndarray is now easily combined with other libraries.

Reviewer 4: Let's link to the newly introduced Interoperability docs.
https://docs-cupy.chainer.org/en/latest/reference/interoperability.html

Author: added

  - DLpack: `ndarray.toDLpack` and `cupy.fromDLpack` can be used to interchange the array with other deep learning frameworks.
  - NumPy: NumPy ufunc is directly applicable to CuPy's ndarray. For example, `numpy.exp(cupy.arange(3))` is valid, which is equivalent to `cupy.exp(cupy.arange(3))`.
  - Numba: Numba's JITed CUDA kernel is directly applicable to CuPy ndarrays.

> We recommend you to update to the latest version of Chainer and CuPy.

Reviewer: ```suggestion
We recommend updating to the latest version of Chainer and CuPy.
```

You can find the upgrade guide [here](https://docs.chainer.org/en/latest/upgrade.html).
Updating Chainer should be done as usual with the command `pip install -U chainer`.
CuPy can be updated in the same way, but be careful to use the appropriate package name if you are using a wheel package (`cupy-cuda

> NN`).

Reviewer 4: Unexpected line break

Author: fixed

> We appreciate any feedback.

Reviewer 3: How about this ( 'would' would be more natural) ?: Any feedback to the dev team would be welcomed and appreciated.

Author: Applied.
